### PR TITLE
feat(curriculum): add weather app lab to cert

### PR DESCRIPTION
**Pull request by majestic-owl448**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #XXXXX

Do I need to add more tests?

### CODE REVIEW

**Comment by ilenia-magoni** [comment]
I need help with this. I don't know how to test the geolocation, nor how to check for error handling

**Comment by a2937** [comment]
I would mock `navigator.geolocation` so that you would be able to test for both behaviors and not be reliant on the user's actual location and browser. 

Let me know if you need any help implementing it.

**Comment by ilenia-magoni** [comment]
> I would mock `navigator.geolocation` so that you would be able to test for both behaviors and not be reliant on the user's actual location and browser.

There are no requirements on function names, so the only thing would be to trigger a DOMContentLoaded event, but looking at documentation it doesn't look like that is possible. I honestly did not try to trigger a DOMContentLoaded event tho so maybe it works 🤔

**Comment by ilenia-magoni** [comment]
TODO: update the app solution to not use the event listener for page loaded, implement a button to click to fetch the data.

ISSUES: the proxy requires lat and long, for an app with the best UI that would need to work with geolocation, which needs to also introduce [the Permissions API](https://developer.mozilla.org/en-US/docs/Web/API/Permissions_API/Using_the_Permissions_API) to check if geolocation is allowed or not. Otherwise change the proxy to accept city name/zip code.

**Comment by a2937** [comment]
Would you feel more comfortable if I updated the weather proxy to use city name and zip?

**Comment by ilenia-magoni** [comment]
If you want, it's probably easier, I don't think we want to add also the Permission API to lecture topics

**Comment by a2937** [comment]
I can do city name.

**Comment by a2937** [comment]
Looking at the [docs](https://openweathermap.org/current#name), while I can do so probably doesn't mean I should do so. Just take the city of Vienna for example. It is the capital of Austria but is also the name of a city in Canada; and the name of several cities in the United States. That being said, I could try to find a way around the issue with your solution and testing.

**Comment by a2937** [comment]
That being said, why not make it so the app is given latitude or longitude ; or an option to use your current location which then uses the permissions API?

**Comment by ilenia-magoni** [comment]
it would need to be a topic added to the lectures, and the idea was already to remove the geolocation API from the lecture topics. the geolocation API needs the permissions API to verify if it's working or not. The API requires latitude and longitude, that is perfect if we use the geolocation API, but if the app is to work with user input, latitude and longitude are not the best UX

**Comment by moT01** [comment]
A PR has been merged that renamed the certification from "Front End" to "Full Stack". This PR may need an update and a few changes made: [see comment](https://github.com/freeCodeCamp/freeCodeCamp/pull/56694#issuecomment-2444473378)

**Comment by moT01** [comment]
A PR has been merged that moves the block order from the `meta.json` file to the `full-stack.json` file.](https://github.com/freeCodeCamp/freeCodeCamp/pull/56795) This PR will need an update, and the order will need to be removed from the `meta.json` file and moved to the new file.

File: curriculum/challenges/english/25-front-end-development/lab-weather-app/66f12a88741aeb16b9246c59.md [hunk 1]
```diff
@@ -0,0 +1,461 @@
...
+assert.exists(document.querySelector("#weather-main"));
+```
+
+You should have an element with the id `location` element for displaying the current location.
+
+```js
+assert.exists(document.querySelector("#location"));
+```
+
+You should have a `clickHandler` function.
```
Comment (Dario-DC) [inline curriculum/challenges/english/25-front-end-development/lab-weather-app/66f12a88741aeb16b9246c59.md:89]: Not mentioned before

File: curriculum/challenges/english/25-front-end-development/lab-weather-app/66f12a88741aeb16b9246c59.md [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+
+```js
+assert.isFunction(clickHandler);
+```
+
+You should have a `checkPermissions` function.
```
Comment (Dario-DC) [inline curriculum/challenges/english/25-front-end-development/lab-weather-app/66f12a88741aeb16b9246c59.md:95]: Not mentioned before

File: curriculum/challenges/english/25-front-end-development/lab-weather-app/66f12a88741aeb16b9246c59.md [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+
+```js
+assert.isFunction(checkPermissions);
+```
+
+You should have a `showWeather` function.
```
Comment (Dario-DC) [inline curriculum/challenges/english/25-front-end-development/lab-weather-app/66f12a88741aeb16b9246c59.md:101]: Not mentioned before

File: curriculum/challenges/english/25-front-end-development/lab-weather-app/66f12a88741aeb16b9246c59.md [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+
+```js
+assert.isFunction(showWeather);
+```
+
+You should have a `getWeather` function.
+
+```js
+assert.isFunction(getWeather);
```
Comment (Dario-DC) [inline curriculum/challenges/english/25-front-end-development/lab-weather-app/66f12a88741aeb16b9246c59.md:110]: It seems that it doesn't work with async functions :(

File: curriculum/challenges/english/25-front-end-development/lab-weather-app/66f12a88741aeb16b9246c59.md [hunk 5]
```diff
@@ -0,0 +1,525 @@
...
+const temp1 = fetch;
+const temp2 = console.log;
+const temp3 = console.error;
+async () => {
+  try {
+    console.log = obj => {testArr.push(obj.toString())};
+    console.error = obj => {testArr.push(obj.toString())};
+    fetch = source => {throw new Error("This is a test error");}
+    await getWeather("chicago");
+    assert.deepEqual(testArr, ["Error: This is a test error"]);
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/25-front-end-development/lab-weather-app/66f12a88741aeb16b9246c59.md:293]:
```suggestion
    assert.include(testArr[0], "This is a test error");
    assert.lengthOf(testArr, 1);
```

**Review by moT01 — COMMENTED** [review]
I think I had a little too much fun with this one @ilenia-magoni 😃 One of my thoughts was that the demo could use a little more style - I ended up going a little overboard with my project and I'm hoping you will use it as the demo/solution...

<details><summary>Here's the code</summary>

````

```html
<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <title>Weather App</title>
    <link rel="stylesheet" href="styles.css">
</head>

<body>
    <div class="weather-info-wrap">
        <div id="weather-info">
            <div id="location"></div>
            <div class="primary-info">
                <div class="primary-info-left" id="main-temperature"></div>
                <div class="primary-info-right">
                    <img id="weather-icon" >
                    <div id="weather-main"></div>
                </div>
            </div>
            <hr >
            <div class="secondary-info">
                <div class="secondary-info-top">
                    <div id="humidity"></div>
                    <div id="feels-like"></div>
                </div>
                <div class="secondary-info-bottom">
                    <div class="secondary-info-bottom-left">
                        <div id="wind"></div>
                        <div id="wind-gust"></div>
                    </div>
                    <div class="secondary-info-bottom-right" id="compass">
                        <span id="compass-arrow">
                            <span class="arrow-head"></span>
                        </span>
                        <span class="compass-direction north"></span>
                        <span class="compass-direction east"></span>
                        <span class="compass-direction south"></span>
                        <span class="compass-direction west"></span>
                    </div>
                </div>
            </div>
        </div>
    </div>

    <div class="btn-wrap">
        <span>Weather for:</span>
        <select id="location-selector">
            <option value=""></option>
            <option value="new york">New York</option>
            <option value="los angeles">Los Angeles</option>
            <option value="chicago">Chicago</option>
            <option value="paris">Paris</option>
            <option value="tokyo">Tokyo</option>
            <option value="london">London</option>
        </select>
        <button id="get-forecast">Get Forecast</button>
    </div>

    <script type="text/javascript" src="script.js"></script>
</body>

</html>
```

```css
* {
  box-sizing: border-box;
}

body {
  background-color: #32a852;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  row-gap: 40px;
  font-size: 20px;
  font-family: sans-serif;
}

.btn-wrap {
  min-width: 700px;
  display: flex;
  justify-content: space-evenly;
  column-gap: 20px;
  border: 2px solid black;
  padding: 20px;
  border-radius: 10px;
  background-color: #6cd488;
  box-shadow: 6px 6px 6px rgba(0,0,0,0.5);
}

.btn-wrap > span {
  text-transform: uppercase;
  text-align: center;
  margin: auto;
}

.btn-wrap > span::first-letter {
  font-size: 1.3em
}

#location-selector {
  width: 200px;
  height: 50px;
  font-size: 20px;
}

#get-forecast {
  font-size: 20px;
  height: 50px;
  width: 200px;
}

.weather-info-wrap {
  min-width: 700px;
  max-width: 700px;
  min-height: 427px;
  padding: 20px 0px;
  text-align: center;
  background-color: #eee;
  border-radius: 10px;
  border: 2px solid black;
  box-shadow: 6px 6px 6px rgba(0,0,0,0.5);
}

#weather-info {
  display: none;
  font-family: sans-serif;
  padding: 20px 0;
}

#location {
  font-size: 30px;
}

.primary-info {
  display: flex;
  justify-content: space-evenly;
  margin: 20px;
  font-size: 25px;
}

.primary-info-left {
  display: flex;
  align-items: center;
}

.primary-info-right {
  display: flex;
  justify-content: center;
  align-items: center;
  column-gap: 10px;
}

.secondary-info {
  margin: 40px 20px 0 20px;
  display: flex;
  flex-direction: column;
  row-gap: 20px;
}

.secondary-info-top {
  display: flex;
  align-items: center;
  justify-content: space-evenly;
}

.secondary-info-bottom {
  display: flex;
  align-items: center;
  justify-content: space-evenly;
  margin-top: 40px;
}

.secondary-info-bottom-left {
  display: flex;
  align-items: center;
  flex-direction: column;
  row-gap: 20px;
  justify-content: space-evenly;
}

.secondary-info-bottom-right {
  width: 100px;
  height: 100px;
  border: 2px solid black;
  border-radius: 50%;
  position: relative;
}

#compass-arrow {
  width: 3px;
  height: 60px;
  position: absolute;
  background-color: black;
}

#compass-arrow {
  width: 4px;
  height: 70px;
  top: 15%;
  position: absolute;
  background-color: black;
}

.arrow-head {
  width: 0;
  height: 0;
  border: 8px solid transparent;
  border-left-color: black;
  position: absolute;
  top: -15%;
  left: -6px;
  transform: rotate(-90deg);
}

.compass-direction {
  position: absolute;
  background-color: black;
}

.north, .south {
  width: 3px;
  height: 8px;
  left: 50%;
}

.east, .west {
  width: 8px;
  height: 3px;
  top: 50%;
}

.north {
  top: -8px;
}

.south {
  bottom: -8px;
}

.east {
  right: -8px;
}

.west {
  left: -8px;
}
```

```js
const getForecastBtn = document.getElementById('get-forecast');
const selectEl = document.getElementById('location-selector');

const weatherInfoEl = document.getElementById('weather-info');
const iconEl = document.getElementById('weather-icon');
const tempEl = document.getElementById('main-temperature');
const feelEl = document.getElementById('feels-like');
const humidityEl = document.getElementById('humidity');
const windEl = document.getElementById('wind');
const gustEl = document.getElementById('wind-gust');
const mainEl = document.getElementById('weather-main');
const locationEl = document.getElementById('location');
const arrowEl = document.getElementById('compass-arrow');

getForecastBtn.addEventListener('click', () =>
  selectEl.value && showWeather(selectEl.value)
);

async function showWeather(city) {
  const json = await getWeather(city);
  const {
    weather,
    main:
    { 
      temp,
      feels_like,
      humidity
    },
    wind: { speed, gust, deg = 0 },
    name
  } = json;

  const { main, icon } = weather[0];

  weatherInfoEl.style.display = 'block';
  
  iconEl.src = icon || '';
  tempEl.innerHTML = temp ? `${temp}&deg; C` : 'N/A';
  feelEl.innerHTML = `Feels Like: ${feels_like ? `${feels_like}&deg; C` : 'N/A'}`;
  humidityEl.innerHTML = `Humidity: ${humidity ? `${humidity}%` : 'N/A'}`;
  windEl.innerHTML = `Wind: ${speed ? `${speed} m/s` : 'N/A'}`;
  gustEl.innerHTML = `Gusts: ${gust ? `${gust} m/s` : 'N/A'}`;

  mainEl.innerHTML = main || 'N/A';
  locationEl.innerHTML = name || 'N/A';
  arrowEl.style.transform = `rotate(${deg}deg)`;
}

async function getWeather(city) {
  try {
    const response = await fetch(`https://weather-proxy.freecodecamp.rocks/api/city/${city}`)

    if (!response.ok) {
      alert('Something went wrong, please try again later');
    }

    const json = await response.json();
    return json;
  } catch (error) {
    console.error(error.message);
  }
}
```

````

</details>

<details><summary>And a screen shot</summary>

Screenshot 2024-12-05 at 11 55 52 AM

</details>

Feel free to make changes. The only test it doesn't pass is the `13. You should have a clickHandler function.` one - which I think can be deleted - it doesn't seem necessary. I left few other comments to try and clarify some of the text.

And here's a few other things I wrote down...

I would add async to the `showWeather` user story. e.g. "You should have an asynchronous function named `showWeather`..."

We could potentially add an example of the JSON that gets returned from the API so users don't have to look at the logs. Not necessary though.

That's about it - it's looking real good 👍

File: curriculum/challenges/english/25-front-end-development/lab-weather-app/66f12a88741aeb16b9246c59.md [hunk 6]
```diff
@@ -0,0 +1,526 @@
...
+       assert.exists(document.querySelector("#location"));
+}
+```
+
+You should have a `clickHandler` function.
+
+```js
+assert.isFunction(clickHandler);
+```
+
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/25-front-end-development/lab-weather-app/66f12a88741aeb16b9246c59.md:183]:
```suggestion
```

File: curriculum/challenges/english/25-front-end-development/lab-weather-app/66f12a88741aeb16b9246c59.md [hunk 7]
```diff
@@ -0,0 +1,526 @@
...
+1. You should handle any errors that occur within the `getWeather` function and log them to the console.
+
+1. You should have a `showWeather` function that accepts a city as parameter.
+
+1. The `showWeather` function should call the `getWeather` function to retrieve the weather data for a certain city.
+
+1. If the `getWeather` function had an error and returns a falsy value, the `showWeather` function should only show an alert that says `Something went wrong, please try again later`.
+
+1. If the data from `getWeather` are usable, the `showWeather` function should display the weather data in the corresponding elements.
```
Comment (ilenia-magoni) [inline curriculum/challenges/english/25-front-end-development/lab-weather-app/66f12a88741aeb16b9246c59.md:57]:
```suggestion
1. If the data from `getWeather` are usable, the `showWeather` function should display the weather data in the corresponding elements. If a certain value from the API is `undefined`, you should write `N/A` in the corresponsing element.
```

**Review by Dario-DC — COMMENTED** [review]
Just tested and it looks great. I think it would be helpful to show an example of the data returned by the API.

Is this really necessary: `If the value from the API is undefined, you should show N/A.` ? Because I didn't use the `N/A` for undefined values and I was able to pass anyway.

Also, tests take a long time to run and I think we should add some kind of note to let the users know. Initially I thought there were some issues with the environment.

File: curriculum/challenges/english/25-front-end-development/lab-weather-app/66f12a88741aeb16b9246c59.md [hunk 8]
```diff
@@ -0,0 +1,876 @@
...
+challengeType: 14 
+dashedName: lab-weather-app
+demoType: onClick
+---
+
+# --description--
+
+**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab. 
+
+You will use a weather API, the data you need is accessible like in this JSON example:
```
Comment (Dario-DC) [inline curriculum/challenges/english/25-front-end-development/lab-weather-app/66f12a88741aeb16b9246c59.md:13]:
```suggestion
You will use a weather API. The output data has the following format:
```

**Review by Dario-DC — APPROVED** [review]

**Review by moT01 — APPROVED** [review]
LGTM 🎉 

The only other thing that came to mind just now is that the weather from the API is static and not current. We could maybe mention that, but I think it's fine for now 🚢

**Pull request merged**